src/mmds/src/token.rs

```diff
@@ -0,0 +1,371 @@
+                Please provide a value between {} and {}.",
+                value, MIN_TOKEN_TTL_SECONDS, MAX_TOKEN_TTL_SECONDS,
+            ),
+            TokenError::ReadFromEntropyPool(err) => {
+                write!(f, "Failed to read from entropy pool: {}.", err)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum TokenStoreError {
+    /// Failed to generate session token.
+    TokenGenerationError(TokenError),
+    /// Session tokens limit exceeded.
+    TokensLimitExceeded,
+}
+
+impl fmt::Display for TokenStoreError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &*self {
+            TokenStoreError::TokenGenerationError(err) => {
+                write!(f, "Failed to generate session token: {}", err)
+            }
+            TokenStoreError::TokensLimitExceeded => {
+                write!(f, "Tokens limit exceeded: {}.", MAX_SESSIONS)
+            }
+        }
+    }
+}
+
+/// Structure for information regarding token lifetime.
+#[derive(Clone)]
+pub struct TokenInfo {
+    ttl_seconds: u32,
+    start_time: DateTime<Utc>,
+}
+
+impl TokenInfo {
+    /// Create a new session Token for MMDS requests.
+    fn new(ttl_seconds: u32) -> Self {
+        TokenInfo {
+            ttl_seconds,
+            start_time: Utc::now(),
+        }
+    }
+
+    /// Return the expiry date of the token.
+    fn expiry(&self) -> DateTime<Utc> {
+        self.start_time
+            .checked_add_signed(Duration::seconds(self.ttl_seconds as i64))
+            .expect("Failed to compute token expiry date.")
+    }
+
+    /// Specifies whether the token has expired.
+    pub fn has_expired(&self) -> bool {
+        Utc::now()
+            .signed_duration_since(self.expiry())
+            .num_milliseconds()
+            >= 0
+    }
+
+    /// Validate the token time to live against bounds.
+    pub fn check_ttl(ttl: u32) -> bool {
+        (MIN_TOKEN_TTL_SECONDS..=MAX_TOKEN_TTL_SECONDS).contains(&ttl)
+    }
+}
+
+/// Structure for the Token used by IMDSv2.
+#[derive(Clone)]
+pub struct Token {
+    value: String,
+    token_info: TokenInfo,
+}
+
+impl Token {
+    /// Create a new session Token for MMDS requests.
+    fn new(value: String, token_info: TokenInfo) -> Self {
+        Token { value, token_info }
+    }
+
+    /// Generate a new random token with the specified lifetime.
+    pub fn generate(ttl_seconds: u32) -> Result<Self, TokenError> {
+        // Validate token time to live.
+        if !TokenInfo::check_ttl(ttl_seconds) {
+            return Err(TokenError::InvalidTtlValue(ttl_seconds));
+        }
+
+        let mut random_bytes = [0u8];
+        let mut token: String = String::new();
+        let mut random_file =
+            File::open(Path::new(RANDOMNESS_POOL)).map_err(TokenError::EntropyPool)?;
+
+        while token.len() < TOKEN_LEN {
+            random_file
+                .read(&mut random_bytes)
+                .map_err(TokenError::ReadFromEntropyPool)?;
```

Comment:
It's inefficient to read one character at a time, try reading blocks of `TOKEN_LEN` and filter out non-alphanumeric.

Reply:
Done